### ISSUE TITLE
Cannot chain async events from non-async events

### ISSUE THREAD

**Issue by masenf** [pynecone-io/pynecone#921]
**Describe the bug**

When an event handler returns an event handler, it doesn't work in the current `main` branch:

```
ValueError: Invalid path: ['State']
```

Same error is seen whether the chained event handler is `async` or not.

When an event handler calls an `async` event handler without `await`, there's a validation error

```
pydantic.error_wrappers.ValidationError: 1 validation error for EventHandler
fn
  <coroutine object State.c2 at 0x1138ba190> is not callable (type=type_error.callable; value=<coroutine object State.c2 at 0x1138ba190>)
/opt/homebrew/Cellar/python@3.11/3.11.3/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py:1907: RuntimeWarning:

coroutine 'State.c2' was never awaited
```

When an event handler `await`s another `async` handler, the state update is NOT sent out while the chained event is being processed.

**To Reproduce**

```python
import asyncio

import pynecone as pc


class State(pc.State):
    val: str = "unset"

    def c1a(self):
        self.val = "LOADING c1a"
        return self.c2

    def c1b(self):
        self.val = "LOADING c1b"
        return self.c2("foo")

    def c1c(self):
        self.val = "LOADING c1c"
        return self.c2()

    async def c1d(self):
        self.val = "LOADING c1d"
        return await self.c2()

    async def c1e(self):
        self.val = "LOADING c1e"
        return await self.c2("foo")

    async def c2(self, val="bar"):
        await asyncio.sleep(0.2)
        self.val = val


def index() -> pc.Component:
    return pc.center(
        pc.vstack(
            pc.heading(State.val, font_size="2em"),
            pc.responsive_grid(
                pc.button("c1a: 'bar' (return)", on_click=State.c1a),
                pc.button("c1b: 'foo' (call w/ args)", on_click=State.c1b),
                pc.button("c1c: 'bar' (call no args)", on_click=State.c1c),
                pc.button("c1d: 'bar' (await no args)", on_click=State.c1d),
                pc.button("c1e: 'foo' (await w/ args)", on_click=State.c1e),
                columns=[2],
                spacing="5",
            ),
        ),
        padding_top="10%",
    )


# Add state and page to the app.
app = pc.App(state=State)
app.add_page(index)
app.compile()
```

**Expected behavior**

**c1a**

Should briefly show the "LOADING c1a" message, then show "bar". **This works on v0.1.27**

Instead:

```console
Task exception was never retrieved
future: <Task finished name='Task-55' coro=<AsyncServer._handle_event_internal() done, defined at /Users/masen/code/pynecone-io/VENV-dev/lib/python3.11/site-packages/socketio/asyncio_server.py:522> exception=ValueError("Invalid path: ['State']")>
Traceback (most recent call last):
  File "/Users/masen/code/pynecone-io/VENV-dev/lib/python3.11/site-packages/socketio/asyncio_server.py", line 524, in _handle_event_internal
    r = await server._trigger_event(data[0], namespace, sid, *data[1:])
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/masen/code/pynecone-io/VENV-dev/lib/python3.11/site-packages/socketio/asyncio_server.py", line 569, in _trigger_event
    return await self.namespace_handlers[namespace].trigger_event(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/masen/code/pynecone-io/VENV-dev/lib/python3.11/site-packages/socketio/asyncio_namespace.py", line 37, in trigger_event
    ret = await handler(*args)
          ^^^^^^^^^^^^^^^^^^^^
  File "/Users/masen/code/pynecone-io/pynecone/pynecone/app.py", line 601, in on_event
    updates = await process(self.app, event, sid, headers, client_ip)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/masen/code/pynecone-io/pynecone/pynecone/app.py", line 454, in process
    updates = [await state._process(event)]
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/masen/code/pynecone-io/pynecone/pynecone/state.py", line 623, in _process
    substate = self.get_substate(path)
               ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/masen/code/pynecone-io/pynecone/pynecone/state.py", line 605, in get_substate
    raise ValueError(f"Invalid path: {path}")
ValueError: Invalid path: ['State']
```

**c1b** and **c1c**

Should briefly show the "LOADING c1b", then show "foo".

Should briefly show the "LOADING c1c", then show "bar".

Instead, both raise error:

```console
Task exception was never retrieved
future: <Task finished name='Task-100' coro=<AsyncServer._handle_event_internal() done, defined at /Users/masen/code/pynecone-io/VENV-dev/lib/python3.11/site-packages/socketio/asyncio_server.py:522> exception=ValidationError(model='EventHandler', errors=[{'loc': ('fn',), 'msg': '<coroutine object State.c2 at 0x10b5ea0c0> is not callable', 'type': 'type_error.callable', 'ctx': {'value': <coroutine object State.c2 at 0x10b5ea0c0>}}])>
Traceback (most recent call last):
  File "/Users/masen/code/pynecone-io/VENV-dev/lib/python3.11/site-packages/socketio/asyncio_server.py", line 524, in _handle_event_internal
    r = await server._trigger_event(data[0], namespace, sid, *data[1:])
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/masen/code/pynecone-io/VENV-dev/lib/python3.11/site-packages/socketio/asyncio_server.py", line 569, in _trigger_event
    return await self.namespace_handlers[namespace].trigger_event(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/masen/code/pynecone-io/VENV-dev/lib/python3.11/site-packages/socketio/asyncio_namespace.py", line 37, in trigger_event
    ret = await handler(*args)
          ^^^^^^^^^^^^^^^^^^^^
  File "/Users/masen/code/pynecone-io/pynecone/pynecone/app.py", line 601, in on_event
    updates = await process(self.app, event, sid, headers, client_ip)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/masen/code/pynecone-io/pynecone/pynecone/app.py", line 454, in process
    updates = [await state._process(event)]
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/masen/code/pynecone-io/pynecone/pynecone/state.py", line 631, in _process
    return await self._process_event(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/masen/code/pynecone-io/pynecone/pynecone/state.py", line 667, in _process_event
    events = fix_events(events, token)
             ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/masen/code/pynecone-io/pynecone/pynecone/event.py", line 366, in fix_events
    e = EventHandler(fn=e)
        ^^^^^^^^^^^^^^^^^^
  File "pydantic/main.py", line 342, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for EventHandler
fn
  <coroutine object State.c2 at 0x10b5ea0c0> is not callable (type=type_error.callable; value=<coroutine object State.c2 at 0x10b5ea0c0>)
/opt/homebrew/Cellar/python@3.11/3.11.3/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py:1907: RuntimeWarning:

coroutine 'State.c2' was never awaited
```

**c1d** and **c1e**

Should briefly show the "LOADING c1d" message, then show "bar".

Should briefly show the "LOADING c1e" message, then show "foo".

The problem here is that the `await` inside the handler itself blocks the initial delta from being sent to the client, so the "LOADING ..." message is not displayed.

** Specifics (please complete the following information):**
 - Python Version: 3.11.2
 - Pynecone Version: 0.1.27
 - OS: macOS 13.2.1
 - Browser (Optional):
 
**Additional context**
Not sure if the intent was to have a behavior change here, but currently the documentation and a lot of my existing code is breaking on `main` due to the inability to return an EventHandler from an EventHandler... `EventHandlers` now seem to be `functools.partial` objects that do not chain correctly.

c1b - c1e do NOT work on v0.1.27 because that version doesn't support EventHandler chaining with args (as far as I can tell)

This issue is related to https://github.com/pynecone-io/pcweb/issues/62, but it's not clear if just a doc change is needed or if there's an actual bug to fix here.

Personally, I think it should still be possible to return an EventHandler (without calling it), and retain the semantics of sending the current delta, and then processing the chained event. I also think it would be nice to return handlers with chained args, that get processed after sending the current delta.

**Comment by sheldonchiu** [comment]
I also encountered this "coroutine xxx was never awaited" issue after upgrading to the latest commit. 

Not sure why, but deleting the .web and all pycache folder, then rebuild with pc init and pc run fixed this issue for me.

**Comment by sheldonchiu** [comment]
for anyone that needs this function, below is my workaround for case c1b (works with latest commit)
```python
from pynecone.event import EventHandler
    def c1b(self):
        self.val = "LOADING c1b"
        return EventHandler(fn=self.c2.func)("foo")
```

**Comment by Lendemor** [comment]
Here is the correct syntax for chaining events (either with non-async or with async events): 
```
class State(pc.State)
    val: str = "unset"
    def c1a(self):
        self.val = "LOADING c1a"
        return State.c2

    def c1b(self):
        self.val = "LOADING c1b"
        return State.c2("foo")

    def c1c(self):
        self.val = "LOADING c1c"
        return State.c2()

    async def c1d(self):
        self.val = "LOADING c1d"
        return State.c2()

    async def c1e(self):
        self.val = "LOADING c1e"
        return State.c2("foo")

    async def c2(self, val="bar"):
        await asyncio.sleep(2)
        self.val = val
```

**Comment by picklelo** [comment]
Nice catch @Lendemor, closing this as it should be resolved by #876
